### PR TITLE
Add type annotations to most functions

### PR DESCRIPTION
**Pull request by tgoodlet**
This is purely for documentation purposes for now as it should be
obvious a bunch of the signatures aren't using the correct "generics"
syntax (i.e. the use of `(str, int)` instead of `typing.Tuple[str, int])`)
in a bunch of places. We're also not using a type checker yet and besides,
`trio` doesn't really expose a lot of its internal types very well.

cc @vodik

### CODE REVIEW

**Comment by vodik** [comment]
I wouldn't do it like this, imho. Stick to what mypy/typing module wants, even if its more verbose.

Using `dict` and `tuple` directly as annotations throws away useful information you could be encoding about the structure of said types: (e.g. `Tuple[int]`, `Tuple[int, str]`, `Tuple[int, ...]` and `Tuple[...]` all mean different things: tuple of a single int, tuple of a integer and string, tuple that has at least one element of type int, and tuple of any length)

But the fact that this design not only deviates away from community's approach to this problem, but is actively hostile to using `mypy` as a development aid (I run mypy as a background checker in my editor), I have zero interest in supporting it.

I'd personally reject this in any project under my control, so not interested in reviewing it.

**Comment by tgoodlet** [comment]
@vodik changed it all.

I'm just gonna go ahead and try to get `mypy` going in CI in this PR. Not quite there yet but maybe I'll get it running so we can look through the problems together.

**Comment by tgoodlet** [comment]
ping @vodik. Should be running `mypy` clean in CI!

File: tractor/_ipc.py [hunk 1]
```diff
@@ -42,25 +43,25 @@ def __init__(self, stream):
...
         return self._raddr
 
-    async def put(self, data):
+    async def put(self, data: Any) -> int:
         async with self._send_lock:
             return await self.stream.send_all(
                 msgpack.dumps(data, use_bin_type=True))
 
-    async def get(self):
+    async def get(self) -> Any:
```
Comment (vodik) [inline tractor/_ipc.py]: Might want to look and see if you can define a `TypeVar` for this instead: https://mypy.readthedocs.io/en/latest/generics.html

```python
T = TypeVar('T')

class StreamQueue(Generic[T]):
    async def get(self) -> Any:
        pass
```

You can get a little bit more type safety this way if you know what the return type will be situationaly:

```python
queue = StreamQueue[str](steam)
```

For a `StreamQueue` that operates over strings.

File: tractor/_portal.py [hunk 2]
```diff
@@ -51,22 +57,26 @@ class Portal:
...
         self.channel = channel
         # when this is set to a tuple returned from ``_submit()`` then
         # it is expected that ``result()`` will be awaited at some point
         # during the portal's lifetime
         self._result = None
-        self._exc = None
-        self._expect_result = None
+        self._exc: Optional[RemoteActorError] = None
+        self._expect_result: Optional[
+            Tuple[str, Any, str, Dict[str, Any]]
```
Comment (vodik) [inline tractor/_portal.py]: You can do something like this:

```python
Result = Tuple[str, Any, str, Dict[str, Any]

self._expect_result: Optional[Result]
```

File: tractor/_trionics.py [hunk 3]
```diff
@@ -99,19 +105,19 @@ def __init__(self, actor, supervisor=None):
...
-        fn,
-        bind_addr=('127.0.0.1', 0),
-        rpc_module_paths=None,
-        statespace=None,
-        loglevel=None,  # set log level per subactor
+        name: str,
+        fn: typing.Callable,
+        bind_addr: Tuple[str, int] = ('127.0.0.1', 0),
+        rpc_module_paths: List[str] = None,
+        statespace: dict = None,
```
Comment (vodik) [inline tractor/_trionics.py]: Probably meant to fix this one?

File: tractor/_trionics.py [hunk 4]
```diff
@@ -274,13 +287,13 @@ def do_hard_kill(proc):
 
 
 @asynccontextmanager
-async def open_nursery(supervisor=None):
+async def open_nursery() -> typing.AsyncGenerator[None, ActorNursery]:
```
Comment (vodik) [inline tractor/_trionics.py]: Got the arguments backwards? Isn't the second argument the send type? I see you yielding a nursery.

File: tractor/_trionics.py (hunk 3 above)
Comment (tgoodlet) [reply tractor/_trionics.py]: Yup.

**Pull request merged**